### ISSUE TITLE
Add an option to store only local accounts transaction data

### ISSUE THREAD

**Issue by gituser** [ripple/rippled#2129]
Would be nice to have an option to store ONLY local accounts transactions in rippled.

To have a nice full history of particular local accounts (incoming/outgoing) and do not store all other transactions.

Bitcoin is doing exactly this, when you are fully synced and import some address it scans blockchain and adds certain transactions to your wallet.dat. And there is an option called `txindex` it allows you to index all transactions from the blockchain (just like how rippled is currently doing).

**Comment by gituser** [comment]
The main problem why this is relevant is:
when you want to store all history of your local account and set `ledger_history` to full rippled starts to eat storage I/O...

Here is an example:
On the mainnet:

After setting ledger_history to full and removing online_delete rippled rapes I/O on the host.. Constantly reading HDD... I've also set cache to 512mb of the RocksDB and set node_size to large but seems it's not helping at all.

I can't use SSD only HDD for now, because SSDs are already used for something else and I don't want to load them more.

Here is a snip of my config:

[node_size]
large


[ledger_history]
full

# This is primary persistent datastore for rippled.  This includes transaction
# metadata, account states, and ledger headers.  Helpful information can be
# found here: https://ripple.com/wiki/NodeBackEnd
# delete old ledgers while maintaining at least 2000. Do not require an
# external administrative command to initiate deletion.
[node_db]
type=RocksDB
path=/home/ripple/.ripple/db/rocksdb
open_files=2000
filter_bits=12
cache_mb=512
file_size_mb=8
file_size_mult=2
#online_delete=65535
advisory_delete=0

I've set for now ledger_history back to defaults 256 and set online_delete to 65535, seems better.. But that means history of the monitored account will disappear after some time.

Maybe there is a devel patch to track only your own account history or what do I change in ripple code to achieve that?

**Comment by bachase** [comment]
Thanks for the request @gituser.   Unfortunately, I'm not aware of a simple patch to share at this time that achieves what you want.

**Comment by gituser** [comment]
@bachase, thanks for the quick reply.

Yes, unfortunately. The only workaround is to setup websocket over rippled, subscribe to your own account and save that txdata flowing into somewhere else..

**Comment by sublimator** [comment]
Whatever happened to separating the rpc gear from the txn processing/p2p
engine?

**Comment by gituser** [comment]
@sublimator what do you mean by that?

**Comment by sublimator** [comment]
Currently rippled is a monolithic program that does all the txn processing, p2p communications, consensus forming and also keeps  tx history in an sqllite database. There was talk of breaking it up. I'd imagine it would probably be much easier to offload a lot of the rpc / history processing to something implemented in a higher level language.

**Comment by MarkusTeufelberger** [comment]
Full history is several TB of data (about 5 or so at the moment) and you'll have a very hard time to query it without SSD storage. Bitcoin on the other hand has about 0.1 TB of data to work through...

The main bulk is node data by the way, not transaction data (transactions are stored in the transactions.db sqlite database).

I agree that splitting out various parts might be an easier approach, https://github.com/rubblelabs/ripple for example is a relatively lean re-implementation of a lot of non-validation code of rippled in go.

**Comment by gituser** [comment]
Thanks @MarkusTeufelberger. I was waiting for your reply..

Would be nice to have at least an option to store local accounts transactions by default and not prune them when you prune other parts of the database.

**Comment by MarkusTeufelberger** [comment]
Rippled currently has no notion of "local accounts". A workaround would be to just store history for all accounts for the time being, but in any case you need a better storage back-end than HDDs.

As I said, the bigger part of history is node data, I'm not sure if rippled will even serve data it only has in its sqlite databases if it doesn't have the actual nodes of that ledger as well. doing online-delete for individual ledgers is a quite expensive operation, since nodes are reused between ledgers, so to remove only stale ones, one needs to do a LOT of I/O.

Maybe once #2088 lands, RocksDB might be a more viable back-end again even for larger amounts of history. In general rippled requires really high I/O and does caching only internally instead of allowing to attach a redis or memcached server, so it is hard to improve on the I/O front without writing code.

**Comment by gituser** [comment]
>A workaround would be to just store history for all accounts for the time being, but in any case you need a better storage back-end than HDDs.
What's the use of wallet.db file then? 

It seems bitcoin was written much more better than ripple. Why not just copy some basic things from there? E.g. store wallet locally, (encrypt if necessary), store transaction data related to the wallet's accounts.

Implement similar DB backend and add `prune` option to limit overall disk space usage.

**Comment by MarkusTeufelberger** [comment]
Just look inside, it stores mostly node identity related stuff, not user keys or anything related to signing transactions on RCL.

Rippled is not based on bitcoin code, so it would be kinda hard to "copy over some basic things".

Again, rippled does not have any idea what "wallets" even are, it is designed to store, process and serve all data of the network. Your features would be better suited for an actual wallet implementation than for a general network node like rippled.

Instead of pruning, rippled has the concept of online_delete since it is not UTXO based and also carries a fully hashed version of its complete state at every ledger close, which makes operating it more IO intense but allows it to drop nearly all historic data at any point in time and still remain functional.

Bitcoin-core's database backend is a slower version of RocksDB combined with simply dumping serialized blocks to disk, it already lead to a hard fork in their block chain and as you said they are processing a fraction of much simpler transactions per second compared to rippled (also their workload is far more bursty in nature due to their early design choices).

Last but not least, also Bitcoin implementations are moving towards this model, https://github.com/btcsuite/btcd is a good example or https://github.com/spesmilo/electrum-server - bitcoin-core is of course moving more at glacial pace, but also tries hard to de-couple wallet code from actual operations and consensus code.

**Comment by gituser** [comment]
@MarkusTeufelberger thanks for the clarifications.

For now I'm gonna use as a workaround - https://data.ripple.com/, because I can't afford running separate powerful server just for rippled with full ledger on RAID array of SSDs. It's not economically viable for me..

**Comment by MarkusTeufelberger** [comment]
No need for RAID, a JBOD should be enough, but yes, running a data API server locally should be a workaround. Maybe it'll even work on HDD.

**Comment by mDuo13** [comment]
I do think it would be great to be able to configure your server to keep "vertical" slices of transaction history based on, say, a set of accounts or other objects to track. The server can follow the `PreviouxTxnID` for those transactions and could probably store only the objects and transactions necessary to reconstruct those transactions and not everything around them.

**Comment by carlhua** [comment]
Should not be implemented by rippled - closing this.
